Brand new arays doc. A small amount of work remains to be done. (#4307)
yugabyte/yugabyte-db: docs/content/latest/api/ysql/datatypes/type_array/literals/_index.md

menu:
  latest:
    identifier: array-literals
    parent: api-ysql-datatypes-array
    weight: 20
isTocNested: false
showAsideToc: false
---

This section introduces array literals informally with a few examples. Its subsections, listed below, explain formally how you construct syntactically correct array literals that establish the values that you intend.

An array literal starts with a left curly brace. This is followed by some number of comma-separated literal representations for the array's values. Sometimes, the value representations need not be double-quoted—but _may_ be. And sometimes the value representations must be double-quoted. The array literal then ends with a right curly brace. Depending on the array's data type, its values might be scalar, or they might be composite. For example, they might be _"row"_ type values; or they might be arrays. The literal for a multidimensional array is written as an array of arrays of arrays... and so on. They might even be values of a user-defined domain which is based on an array data type. This somewhat exotic, but potentially very useful, notion is discussed in a dedicated subsection [here](to_do).

> To use such a literal in SQL or in PL/pgSQL it must be enquoted in the same way as is an ordinary `text` literal. You can enquote an array literal using dollar quotes, if this suits your purpose, just as you can for a `text` literal. You sometimes need to follow the closing quote with a suitable typecast operator for the array datatype that you intend. And sometimes the context of use uniquely determines the literal's datatype. It's never wrong to write the typecast explicitly—and it's a good practice always to do this.

Reviewer: ```suggestion
To use such a literal in SQL or in PL/pgSQL it must be enquoted in the same way as is an ordinary `text` literal. You can enquote an array literal using dollar quotes, if this suits your purpose, just as you can for a `text` literal. You sometimes need to follow the closing quote with a suitable typecast operator for the array data type that you intend. And sometimes the context of use uniquely determines the literal's data type. It's never wrong to write the typecast explicitly—and it's a good practice always to do this.
```

> Here, in use in a SQL `select`, is the literal for a one-dimensional array of primitive `int` values:

Reviewer: ```suggestion
Here, in use in a SQL `SELECT` statement, is the literal for a one-dimensional array of primitive `int` values:
```

```postgresql
\t on
select '{1, 2, 3}'::int[];
```

> The `\t on` metacommand suppresses column headers and the rule-off under these. Unless the headers are important for understanding, query output from _ysqlsh_ will be shown, throughout the present "arrays" major section, without these.

Reviewer: ```suggestion
The `\t on` metacommand suppresses column headers and the rule-off under these. Unless the headers are important for understanding, query output from `ysqlsh` will be shown, throughout the present "arrays" major section, without these.
```

```
It produces the identical output to the first example, where no double quotes were used.

The third example defines a two-dimensional array of `int` values:

```postgresql
select '
   {
      {11, 12, 13},
      {21, 22, 23}
    }
  '::int[];
```

It produces this result:

```
 {{11,12,13},{21,22,23}}
```

The fourth example defines an array whose values are instances of a _"row"_ type:

```postgresql
create type rt as (a int, b text);

select '
  {
    "(1,a1 a2)",
    "(2,b1 b2)",
    "(3,c1 v2)"
  }
'::rt[];
```
It produces this output:
```
 {"(1,\"a1 a2\")","(2,\"b1 b2\")","(3,\"c1 v2\")"}
```
All whitespace (except, of course, within the text values) has been removed. The double quotes around the representation of each _"row"_ type value are retained. This suggests that they are significant. (Test this by removing them. It causes the _"22P02: malformed row literal"_ error.) Most noticeably, there are clearly rules at work in connection with the representation of each `text` value within the representation of each _"row"_ type value.

The following subsections present the rules carefully and, when the rules allow some freedom, give recommendations.

> The [first subsection](./text-typecasting-and-literals/) establishes the important notions that allow you to distinguish between a _literal_ and the _text of the literal_. It's the text of an array literal that, by following specific grammar rules for this class of literal, actually defines the intended value. The literal, as a whole, enquotes this bare text and typecasts it to the desired target array data type.

Reviewer: ```suggestion
First, the [Text typecasting and literals](./text-typecasting-and-literals/) section establishes the important notions that allow you to distinguish between a _literal_ and the _text of the literal_. It's the text of an array literal that, by following specific grammar rules for this class of literal, actually defines the intended value. The literal, as a whole, enquotes this bare text and typecasts it to the desired target array data type.
```

> The [second subsection](./array-of-primitive-values/) gives the rules for array literals whose values are scalars (for example, are of primitive data types).

Reviewer: ```suggestion
Second, the [Array of primitive values](./array-of-primitive-values/) section gives the rules for array literals whose values are scalars (for example, are of primitive data types).
```

> The [third subsection](./row/) gives the rules for the literal for a value of a _"row"_ type. These rules are essential to the understanding of the next section.

Reviewer: ```suggestion
Third, the [Row](./row/) section gives the rules for the literal for a value of a _"row"_ type. These rules are essential to the understanding of the next section.
```

> The [fourth subsection](./array-of-rows/) gives the rules for array literals whose values are composite (that is, a _"row"_ type).

Reviewer: ```suggestion
Fourth, the [Array of rows](./array-of-rows/) section gives the rules for array literals whose values are composite (that is, a _"row"_ type).
```
